akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala

```diff
@@ -29,60 +28,50 @@ import scala.collection.immutable
 /**
  * Represents one direction of an Http2 substream.
  */
-private[http2] sealed trait Http2SubStream {
-  type T
-  val initialHeaders: ParsedHeadersFrame
-  val data: Source[T, Any]
-  val correlationAttributes: Map[AttributeKey[_], _]
+@InternalApi
+private[http2] case class Http2SubStream(
+  initialHeaders:        ParsedHeadersFrame,
+  data:                  Source[Any, Any],
+  correlationAttributes: Map[AttributeKey[_], _]
+) {
   def streamId: Int = initialHeaders.streamId
-
-  def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream
-
-  def createRequestEntity(contentLength: Long, contentType: ContentType): RequestEntity =
-    this match {
-      case s if s.data == Source.empty || contentLength == 0 => HttpEntity.Empty
-      case s: ByteHttp2SubStream =>
-        if (contentLength > 0) HttpEntity.Default(contentType, contentLength, s.data)
-        else HttpEntity.Chunked.fromData(contentType, s.data)
-      /* contentLength undefined */
-      case c: ChunkedHttp2SubStream => HttpEntity.Chunked(contentType, c.data)
+  def hasEntity: Boolean = !initialHeaders.endStream
+
+  def withCorrelationAttributes(newAttributes: Map[AttributeKey[_], _]): Http2SubStream =
+    copy(correlationAttributes = newAttributes)
+
+  def createEntity(contentLength: Long, contentTypeOption: OptionVal[ContentType]): RequestEntity = {
```

Comment:
```suggestion
  def createEntity(contentLength: Long, contentTypeOption: OptionVal[ContentType]): ResponseEntity = {
```

Reply:
`RequestEntity <: ResponseEntity`, so `RequestEntity` is the more concrete type. Why should we change it?

Reply:
Actually it wouldn't compile because it is used for requests as well which have tighter requirements.

Reply:
hm, not sure what I was thinking :laughing: